core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+            for (RemoteSourceNode remoteSource : stage.getFragment().getRemoteSourceNodes()) {
+                ExchangeSourceOutputSelector selector = null;
+                for (PlanFragmentId sourceFragment : remoteSource.getSourceFragmentIds()) {
+                    ExchangeSourceOutputSelector fragmentSelector = sourceOutputSelectors.get(sourceFragment);
+                    if (fragmentSelector == null) {
+                        continue;
+                    }
+                    if (selector == null) {
+                        selector = fragmentSelector;
+                    }
+                    else {
+                        selector = selector.merge(fragmentSelector);
+                    }
+                    if (selector != null) {
+                        partition.updateExchangeSourceOutputSelector(remoteSource.getId(), selector);
+                    }
+                }
+            }
+            remainingPartitions.add(partitionId);
+
+            return Optional.of(new ScheduledTask(stage.getStageId(), partitionId, SPECULATIVE_EXECUTION_PRIORITY + schedulingPriority));
+        }
+
+        public void updatePartition(int partitionId, PlanNodeId planNodeId, List<Split> splits, boolean noMoreSplits)
+        {
+            if (getState().isDone()) {
+                return;
+            }
+
+            StagePartition partition = getStagePartition(partitionId);
+            partition.updateOpenTaskDescriptor(planNodeId, splits, noMoreSplits);
+        }
+
+        public Optional<ScheduledTask> sealPartition(int partitionId)
+        {
+            if (getState().isDone()) {
+                return Optional.empty();
+            }
+
+            StagePartition partition = getStagePartition(partitionId);
+            TaskDescriptor taskDescriptor = partition.sealTaskDescriptor(partitionId);
+            // a task may finish before task descriptor is sealed
+            if (!partition.isFinished()) {
+                taskDescriptorStorage.put(stage.getStageId(), taskDescriptor);
+            }
+
+            if (!partition.isRunning()) {
+                // if partition is not yet running update it's priority as it is no longer speculative
+                return Optional.of(new ScheduledTask(stage.getStageId(), partitionId, schedulingPriority));
+            }
+
+            // TODO: split into smaller partitions here if necessary (for example if a task for a given partition failed with out of memory)
+
+            return Optional.empty();
+        }
+
+        public void noMorePartitions()
+        {
+            if (getState().isDone()) {
+                return;
+            }
+
+            noMorePartitions = true;
+            if (remainingPartitions.isEmpty()) {
+                stage.finish();
+                // TODO close exchange early
+                taskSource.close();
+            }
+        }
+
+        public Optional<RemoteTask> schedule(int partitionId, InternalNode node)
```

Comment:
I am too dumb to follow this one.

Reply:
Yeah, probably the most beafy method of the entire scheduler. Essentially what it does - it schedules a task.

And to schedule a task you need to have splits and output selectors.

This method obtains splits either from an open descriptor (if a descriptor is still being built) or from a sealed descriptor stored in the task descriptor storage. It also merges in output selectors.

Open to suggestions of how to make it more readable.

Reply:
Actually it is not that bad on the second go. What you could do is to extract: 

```java
        private Set<PlanNodeId> getRemoteSourceIds()
        {
            // this can be cached
            Set<PlanNodeId> remoteSourceIds = new HashSet<>();
            for (RemoteSourceNode remoteSource : stage.getFragment().getRemoteSourceNodes()) {
                remoteSourceIds.add(remoteSource.getId());
            }
            return remoteSourceIds;
        }

        private Map<PlanNodeId, ExchangeSourceOutputSelector> getMergedSourceOutputSelectors()
        {
            ImmutableMap.Builder<PlanNodeId, ExchangeSourceOutputSelector> outputSelectors = ImmutableMap.builder();
            for (RemoteSourceNode remoteSource : stage.getFragment().getRemoteSourceNodes()) {
                ExchangeSourceOutputSelector mergedSelector = null;
                for (PlanFragmentId sourceFragmentId : remoteSource.getSourceFragmentIds()) {
                    ExchangeSourceOutputSelector sourceFragmentSelector = sourceOutputSelectors.get(sourceFragmentId);
                    if (sourceFragmentSelector == null) {
                        continue;
                    }
                    if (mergedSelector == null) {
                        mergedSelector = sourceFragmentSelector;
                    }
                    else {
                        mergedSelector = mergedSelector.merge(sourceFragmentSelector);
                    }
                }
                if (mergedSelector != null) {
                    outputSelectors.put(remoteSource.getId(), mergedSelector);
                }
            }
            return outputSelectors.buildOrThrow();
        }
```

Then you will have just
```java
            Set<PlanNodeId> remoteSourceIds = getRemoteSourceIds();
            Map<PlanNodeId, ExchangeSourceOutputSelector> outputSelectors = getMergedSourceOutputSelectors();
```
in `schedule()`.

Maybe you could also build some common interface adapter over `TaskDescriptor` and `OpenTaskDescriptor` with methods:
```
  public ListMultimap<PlanNodeId, Split> getSplits();
  public boolean wasNoMoreSplits(PlanNodeId remoteSourcePlanNodeId); 
```

then you could mostly unify handling of both. But not sure if that is worth a fuss.

Reply:
Great suggestions. Refactored, it seems to be way more readable now. Please take a look.

Reply:
Thanks